m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -0,0 +1,765 @@
+            bucketValFmt = String.format("%%.%df", builder.histogramBucketTagPrecision);
+
+            metricQueue = new LinkedBlockingQueue<>(builder.maxQueueSize);
+
+            executor = builder.executor;
+
+            addAndRunProcessor(builder.metricTagSet);
+        } catch (TTransportException | SocketException e) {
+            throw new RuntimeException("Exception creating M3Reporter", e);
+        }
+    }
+
+    private int calculateFreeBytes(int maxPacketSizeBytes, Set<MetricTag> commonTags) {
+        MetricBatch metricBatch = new MetricBatch();
+        metricBatch.setCommonTags(commonTags);
+        metricBatch.setMetrics(new ArrayList<Metric>());
+
+        int size;
+
+        try {
+            metricBatch.write(calcProtocol);
+            size = calc.getSizeAndReset();
+        } catch (TException e) {
+            LOG.warn("Unable to calculate metric batch size. Defaulting to: %s", DEFAULT_METRIC_SIZE);
+            size = DEFAULT_METRIC_SIZE;
+        }
+
+        int numOverheadBytes = EMIT_METRIC_BATCH_OVERHEAD + size;
+
+        freeBytes = maxPacketSizeBytes - numOverheadBytes;
+
+        if (freeBytes <= 0) {
+            throw new IllegalArgumentException("Common tags serialized size exceeds packet size");
+        }
+
+        return freeBytes;
+    }
+
+    private static String getHostName() {
+        try {
+            return InetAddress.getLocalHost().getHostName();
+        } catch (UnknownHostException e) {
+            LOG.warn("Unable to determine hostname. Defaulting to: %s", DEFAULT_TAG_VALUE);
+            return DEFAULT_TAG_VALUE;
+        }
+    }
+
+    private void addAndRunProcessor(Set<MetricTag> commonTags) {
+        phaser.register();
+
+        executor.execute(new Processor(commonTags));
+    }
+
+    @Override
+    public Capabilities capabilities() {
+        return CapableOf.REPORTING_TAGGING;
+    }
+
+    @Override
+    public void flush() {
+        try {
+            metricQueue.put(SizedMetric.FLUSH);
+        } catch (InterruptedException e) {
+            LOG.warn("Interrupted while trying to queue flush sentinel");
+        }
+    }
+
+    private void flush(List<Metric> metrics, Set<MetricTag> commonTags) {
+        flush(metrics, commonTags, DEFAULT_FLUSH_TRY_COUNT);
+    }
+
+    private void flush(List<Metric> metrics, Set<MetricTag> commonTags, int maxTries) {
+        if (metrics.isEmpty()) {
+            return;
+        }
+
+        int tries = 0;
+        boolean tryFlush = true;
+
+        MetricBatch metricBatch = new MetricBatch();
+        metricBatch.setCommonTags(commonTags);
+        metricBatch.setMetrics(metrics);
+
+        while (tryFlush) {
+            try {
+                client.emitMetricBatch(metricBatch);
+
+                tryFlush = false;
+            } catch (TException tException) {
+                tries++;
+
+                if (tries < maxTries) {
+                    try {
+                        Thread.sleep(FLUSH_RETRY_DELAY_MILLIS);
```

Comment:
why sleep after a failure?

Reply:
Agreed that this isn't the prettiest code, but I want to sleep only when I hit an exception, and sleep throws a checked exception 😕. Extracting out just a sleep function seems a little silly.

Reply:
As discussed offline, there seems to be little upside to having this retry logic in the first place - removing this for now.